doc/manual/src/language/string-interpolation.md

```diff
@@ -0,0 +1,97 @@
+# String interpolation
+
+String interpolation is a language feature where a [string], [path], or [attribute name] can contain expressions enclosed in `${ }` (dollar-sign with curly brackets).
+
+Such a string is an *interpolated string*, and an expression inside is an *interpolated expression*.
+
+Interpolated expressions must evaluate to something that can be coerced to a string:
+
+- [string]
+- [path]
+- [derivation]
+
+[string]: ./language/values.md#type-string
+[path]: ./language/values.md#type-path
+[attribute name]: ./language/values.md#attribute-set
+[derivation]: ../glossary.md#gloss-derivation
+
+When an interpolated expression evaluates to a path, and is thus coerced into a string, the path is first copied into the Nix store and the resulting string is the [store path].
+
+For instance, evaluating `"${./foo.txt}" will cause `foo.txt` in the current directory to be copied into the Nix store and result in the string `"/nix/store/<hash>-foo.txt"`.
+
+[store path]: ../glossary.md#gloss-store-path
+
+Note that the Nix language assumes that all input files will remain _unchanged_ during the course of the Nix expression evaluation.
+
+For example, assume you used a file path in an interpolated string during a `nix repl` session.
+Later in the same session, after having changed the file contents, evaluating the interpolated string with the file path again will still return the first store path.
+Nix will _not_ re-read the file contents to produce a different store path.
+
+## Examples
+
+### String
+
+Rather than writing
+
+```nix
+"--with-freetype2-library=" + freetype + "/lib"
+```
+
+(where `freetype` is a [derivation]), you can instead write
+
+```nix
+"--with-freetype2-library=${freetype}/lib"
+```
+
+The latter is automatically translated to the former.
+
+A more complicated example (from the Nix expression for [Qt](http://www.trolltech.com/products/qt)):
+
+```nix
+configureFlags = "
+  -system-zlib -system-libpng -system-libjpeg
+  ${if openglSupport then "-dlopen-opengl
+    -L${mesa}/lib -I${mesa}/include
+    -L${libXmu}/lib -I${libXmu}/include" else ""}
+  ${if threadSupport then "-thread" else "-no-thread"}
+";
+```
+
+Note that Nix expressions and strings can be arbitrarily nested;
+in this case the outer string contains various interpolated expressions that themselves contain strings (e.g., `"-thread"`), some of which in turn contain interpolated expressions (e.g., `${mesa}`).
+
+### Path
+
+Rather than writing
+
+```nix
+./. + "/" + foo + "-" + bar + ".nix"
+```
+
+or
+
+```nix
+./. + "/${foo}-${bar}.nix"
+```
+
+you can instead write
+
+```nix
+./${foo}-${bar}.nix
+```
+
+### Attribute name
+
+Attribute names can be created dynamically with string interpolation:
+
+```nix
+let
+  names = [ "foo" "bar" "baz" ];
+  attrs = map (name: { ${name} = true; }) names;
+in
+builtins.foldl' (x: y: x // y) {} attrs
```

Comment:
I think this is an unnecessarily complex example for someone just trying to understand string interpolation.

Reply:
Agreed. I simply couldn't come up with a simpler one on the spot. How would you demonstrate dynamic attribute names?

Reply:
I think it can be completely logicless, e.g.:

```
let
  key = "keyName";
in {
  ${key} = 42;
}
```

as we're not trying to demonstrate anything else here.